<a href="https://colab.research.google.com/github/Pornphankub/dataviz2025/blob/main/midterm2025.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ให้สืบหาฆาตรกร (10 คะแนน) และ ผู้บงการ (5 คะแนน) ในคดี ฆาตรกรรม (murder) ที่เกิดขึ้นในวันที่ 15 มกราคม 2018 ที่เมือง Pandas City โดยเริ่มจากการตรวจสอบ ข้อมูลคดีในไฟล์ crime_scene_report.csv

In [103]:
import os

In [104]:
import pandas as pd

In [105]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [106]:
path_to_crime_scene_report = '/content/drive/MyDrive/Python&DataViz2025/Midterm/crime_scene_report.csv'

In [107]:
crimescenereport_df = pd.read_csv(path_to_crime_scene_report)
crimescenereport_df.head()

date     type                                        description  \
0  20180115  robbery  A Man Dressed as Spider-Man Is on a Robbery Spree   
1  20180115   murder                  Life? Dont talk to me about life.   
2  20180115   murder  Mama, I killed a man, put a gun against his he...   
3  20180215   murder                         REDACTED REDACTED REDACTED   
4  20180215   murder  Someone killed the guard! He took an arrow to ...   

          city  
0          NYC  
1       Albany  
2         Reno  
3  Pandas City  
4  Pandas City

# หาคดีฆาตกรรมในวันที่ 15 มกราคม 2018

In [108]:
crime_in_date = crimescenereport_df[crimescenereport_df['date'] == 20180115]
crime_in_date

date     type                                        description  \
0     20180115  robbery  A Man Dressed as Spider-Man Is on a Robbery Spree   
1     20180115   murder                  Life? Dont talk to me about life.   
2     20180115   murder  Mama, I killed a man, put a gun against his he...   
5     20180115    theft                    Big Bully stole my lunch money!   
6     20180115    fraud  Lorem ipsum dolor sit amet, consectetur adipis...   
11    20180115  assault  Hamilton: Lee, do you yield? Burr: You shot hi...   
12    20180115  assault                                   Report Not Found   
972   20180115    arson  like a mouse, you know. But do cats eat bats, ...   
973   20180115  bribery  Here the other guinea-pig cheered, and was sup...   
974   20180115    fraud  The Dormouse slowly opened his eyes. ‘I wasn’t...   
975   20180115    fraud  King put on his spectacles and looked anxiousl...   
1227  20180115   murder  Security footage shows that there were 2 witne...   

                  city  
0                  NYC  
1               Albany  
2                 Reno  
5              Chicago  
6              Seattle  
11         Pandas City  
12         Pandas City  
972   San Buenaventura  
973         Marysville  
974             Yakima  
975              Bryan  
1227       Pandas City

In [112]:
murder_in_date = crime_in_date[crime_in_date['type'].str.contains('murder')]
murder_in_date

date    type                                        description  \
1     20180115  murder                  Life? Dont talk to me about life.   
2     20180115  murder  Mama, I killed a man, put a gun against his he...   
1227  20180115  murder  Security footage shows that there were 2 witne...   

             city  
1          Albany  
2            Reno  
1227  Pandas City

In [113]:
murder_in_pandas_city = murder_in_date[murder_in_date['city'] == 'Pandas City']
murder_in_pandas_city

date    type                                        description  \
1227  20180115  murder  Security footage shows that there were 2 witne...   

             city  
1227  Pandas City

In [114]:
if not murder_in_pandas_city.empty:
    for index, row in murder_in_pandas_city.iterrows():
        print(f"--- รายละเอียดข้อมูล: {index} ---")
        for column, value in row.items():
            print(f"{column}: {value}")
else:
    print("DataFrame 'murder_in_pandas_city' is empty.")

--- รายละเอียดข้อมูล: 1227 ---
date: 20180115
type: murder
description: Security footage shows that there were 2 witnesses. The first witness lives at the last house on "Northwestern Dr". The second witness, named Annabel, lives somewhere on "Franklin Ave".
city: Pandas City


สรุปข้อมูล

--- รายละเอียดข้อมูล: 1227 ---

date: 20180115

type: murder

description: ภาพจากกล้องวงจรปิดแสดงให้เห็นว่ามีพยาน 2 คน พยานคนแรกอาศัยอยู่ที่บ้านหลังสุดท้ายบนถนน "Northwestern Dr" พยานคนที่สองชื่อ Annabel อาศัยอยู่ที่ไหนสักแห่งบนถนน "Franklin Ave".

city: Pandas City

## หาข้อมูลพยานคนที่ 1

In [115]:
path_to_person = '/content/drive/MyDrive/Python&DataViz2025/Midterm/person.csv'

In [116]:
person_df = pd.read_csv(path_to_person)
person_df

id                 name  license_id  address_number  \
0      10000   Christoper Peteuil      993845             624   
1      10007  Kourtney Calderwood      861794            2791   
2      10010            Muoi Cary      385336             741   
3      10016          Era Moselle      431897            1987   
4      10025         Trena Hornby      550890             276   
...      ...                  ...         ...             ...   
10006  99936          Luba Benser      274427             680   
10007  99941      Roxana Mckimley      975942            1613   
10008  99965      Cherie Zeimantz      287627            3661   
10009  99982          Allen Cruse      251350            3126   
10010  99990         Vance Hunten      830407            3056   

      address_street_name        ssn  
0            Bankhall Ave  747714076  
1           Gustavus Blvd  477972044  
2         Northwestern Dr  828638512  
3           Wood Glade St  614621061  
4           Daws Hill Way  223877684  
...                   ...        ...  
10006        Carnage Blvd  685095054  
10007             Gate St  512136801  
10008       The Water Ave  362877324  
10009           N Jean Dr  348734531  
10010       Lancefield St  896677562  

[10011 rows x 6 columns]

In [117]:
person1_df = person_df[person_df['address_street_name'] == 'Northwestern Dr']
person1_df

id                  name  license_id  address_number  \
2     10010             Muoi Cary      385336             741   
282   12711        Norman Apolito      667757             599   
499   14887        Morty Schapiro      118009            4919   
528   15171          Weldon Penso      336999             311   
811   17729        Lasonya Wildey      439686            3824   
904   18376              Josh Shi      653712            1091   
1021  19420           Cody Schiel      890431            3524   
1338  22239         Dusty Sigafus      710517            1125   
1426  23044          Val Portlock      924989            3143   
1522  23960  Kristopher Lagerberg      658777            1392   
1686  25615         Olevia Morena      388477             812   
1812  26758      Angelena Billman      942074             629   
1981  28360        Rashad Cascone      924517            3212   
2131  29608       Willetta Lurvey      800568            1028   
2535  33415         Homer Simpson      108392             981   
2620  34155      Allyson Lazenson      814209            2840   
2635  34352    Vincenza Burkhardt      436470            2512   
2871  36378        Freddie Ellzey      267882            3449   
3237  39688            Abe Roeker      537445             689   
3316  40336       Fredia Billeter      251082             404   
3368  40760          Renata Faver      976650             745   
3743  44004           Alison Eska      327797            2951   
3795  44542        Ivory Mcgrevey      873842             897   
4595  51114     Christena Saffell      814284            3055   
4643  51552           Bonita Knie      764124            1726   
4806  53076           Boris Bijou      664914            3327   
4871  53699      Christian Pagano      628433            2585   
4886  53890        Sophie Tiberio      957671            3755   
4956  54564         Sammie Sengun      217286            2590   
5024  55130        Shiloh Huntoon      472118             951   
5450  59307          Cyril Yongue      753524            2903   
6476  68690            Yer Modest      852404            3046   
6698  70485     Andera Kostelecky      354822            1252   
6865  71924        Laurie Scronce      774116            1418   
6877  72076     Courtney Bordeaux      324614            1707   
7040  73368        Torie Thalmann      773862            3697   
7296  75484          Del Tacderen      447887             607   
7888  80921       Donovan Yackley      222395            2386   
7902  81052     Calvin Czelusniak      471483            1516   
7993  81885         Kurtis Masuda      934201            2108   
8369  85280             Donn Duin      569122             540   
8608  87456   Leonora Wolfsberger      215868            3483   
8720  88423           Teri Ehrich      510876            1879   
8888  89906       Kinsey Erickson      510019             309   
8939  90397     Dionna Kranwinkle      797260            2678   
9272  93509        Emmitt Aceuedo      916706            3491   
9569  96178           Galina Khuu      450248            2183   
9609  96595         Coretta Cubie      303645            3631   
9841  98593           Jonah Toner      578982            2947   
9991  99826          Ivy Kazarian      994470             892   

     address_street_name        ssn  
2        Northwestern Dr  828638512  
282      Northwestern Dr  778264744  
499      Northwestern Dr  111564949  
528      Northwestern Dr  131379495  
811      Northwestern Dr  917817122  
904      Northwestern Dr  193899001  
1021     Northwestern Dr  947110049  
1338     Northwestern Dr  724386723  
1426     Northwestern Dr  100593316  
1522     Northwestern Dr  492912529  
1686     Northwestern Dr  962831331  
1812     Northwestern Dr  341854756  
1981     Northwestern Dr  838907424  
2131     Northwestern Dr  843751378  
2535     Northwestern Dr  871392019  
2620     Northwestern Dr  991150469  
2635     Northwestern Dr  197913046  
2871     

In [118]:
# หาหมายเลขที่อยู่สูงสุดบนถนน Northwestern Dr
last_house_number = person1_df['address_number'].max()

# กรองหาคนที่อาศัยอยู่ที่บ้านหลังสุดท้าย
person1_witness = person1_df[person1_df['address_number'] == last_house_number]

# แสดงข้อมูลพยานคนแรก
person1_witness

id            name  license_id  address_number address_street_name  \
499  14887  Morty Schapiro      118009            4919     Northwestern Dr   

           ssn  
499  111564949

## หาพยานคนที่ 2

In [119]:
person2_df = person_df[person_df['address_street_name'] == 'Franklin Ave']
person2_witness = person2_df[person2_df['name'].str.contains('Annabel')]
person2_witness

id            name  license_id  address_number address_street_name  \
665  16371  Annabel Miller      490173             103        Franklin Ave   

           ssn  
665  318771143

## ดูการให้ปากกาคำของพยาน

In [120]:
path_to_intereview = '/content/drive/MyDrive/Python&DataViz2025/Midterm/interview.csv'

In [121]:
interreview_df = pd.read_csv(path_to_intereview)
interreview_df.head(3)

person_id                                     transcript
0      28508            ‘I deny it!’ said the March Hare.\n
1      63713                                             \n
2      86208  way, and the whole party swam to the shore.\n

In [122]:
interreview1 = interreview_df[interreview_df['person_id'] == 14887]

if not interreview1.empty:
    for index, row in interreview1.iterrows():
        print(f"--- Row Index: {index} ---")
        for column, value in row.items():
            print(f"{column}: {value}")
else:
    print("DataFrame 'interreview1' is empty.")

--- Row Index: 4988 ---
person_id: 14887
transcript: I heard a gunshot and then saw a man run out. He had a "Get Fit Now Gym" bag. The membership number on the bag started with "48Z". Only gold members have those bags. The man got into a car with a plate that included "H42W".


--- Row Index: 4988 ---

person_id: 14887

transcript:ฉันได้ยินเสียงปืนแล้วเห็นชายคนหนึ่งวิ่งออกมา เขามีกระเป๋าของ "Get Fit Now Gym" หมายเลขสมาชิกบนกระเป๋าขึ้นต้นด้วย "48Z" เฉพาะสมาชิกระดับ Gold เท่านั้นที่มีกระเป๋าแบบนี้ ชายคนนั้นขึ้นรถยนต์ที่มีป้ายทะเบียนมีส่วนประกอบของ "H42W" อยู่ด้วย

In [123]:
interreview2 = interreview_df[interreview_df['person_id'] == 16371]

if not interreview2.empty:
    for index, row in interreview2.iterrows():
        print(f"--- Row Index: {index} ---")
        for column, value in row.items():
            print(f"{column}: {value}")
else:
    print("DataFrame 'interreview2' is empty.")

--- Row Index: 4989 ---
person_id: 16371
transcript: I saw the murder happen, and I recognized the killer from my gym when I was working out last week on January the 9th.


--- Row Index: 4989 ---

person_id: 16371

transcript:ฉันเห็นเหตุฆาตกรรมเกิดขึ้น และฉันจำฆาตกรได้ว่าเคยเห็นเขาที่ฟิตเนสตอนที่ฉันไปออกกำลังกายเมื่อสัปดาห์ที่แล้ว วันที่ 9 มกราคม


# สรุปข้อมูลที่ใช้หาฆาตกร

- มีกระเป๋าของ "Get Fit Now Gym"
- หมายเลขสมาชิกบนกระเป๋าขึ้นต้นด้วย "48Z"
- เฉพาะสมาชิกระดับ Gold เท่านั้นที่มีกระเป๋าแบบนี้
- ชายคนนั้นขึ้นรถยนต์ที่มีป้ายทะเบียนมีส่วนประกอบของ "H42W"
- วันที่ 9 มกราคม เขาออกไปฟิตเนส

## ตรวจสอบฟิตเนส

In [124]:
path_to_getfitnow_member = '/content/drive/MyDrive/Python&DataViz2025/Midterm/get_fit_now_member.csv'

In [125]:
getfitnow_member = pd.read_csv(path_to_getfitnow_member)
getfitnow_member.head(3)

id  person_id                name  membership_start_date  \
0  NL318      65076     Everette Koepke               20170926   
1  AOE21      39426        Noe Locascio               20171005   
2  2PN28      63823  Jeromy Heitschmidt               20180215   

  membership_status  
0              gold  
1           regular  
2            silver

In [126]:
getfitnow_member_gold = getfitnow_member[getfitnow_member['membership_status'] == 'gold']
getfitnow_member_gold = getfitnow_member_gold[getfitnow_member_gold['id'].str.contains('48Z')]
getfitnow_member_gold

id  person_id           name  membership_start_date membership_status
181  48Z7A      28819   Joe Germuska               20160305              gold
182  48Z55      67318  Jeremy Bowers               20160101              gold

### ตรวจสอบว่าไปฟิตเนสวันที่ 9 มกราไหม

In [127]:
path_to_getfitnow_checkin = '/content/drive/MyDrive/Python&DataViz2025/Midterm/get_fit_now_check_in.csv'

In [128]:
getfitnow_checkin = pd.read_csv(path_to_getfitnow_checkin)
getfitnow_checkin

membership_id  check_in_date  check_in_time  check_out_time
0            NL318       20180212            329             365
1            NL318       20170811            469             920
2            NL318       20180429            506             554
3            NL318       20180128            124             759
4            NL318       20171027            418            1019
...            ...            ...            ...             ...
2698         4KB72       20170422           1016            1114
2699         4KB72       20170630            408             885
2700         48Z7A       20180109           1600            1730
2701         48Z55       20180109           1530            1700
2702         90081       20180109           1600            1700

[2703 rows x 4 columns]

In [129]:
filtered_checkins = getfitnow_checkin[
    getfitnow_checkin['membership_id'].isin(['48Z7A', '48Z55']) &
    (getfitnow_checkin['check_in_date'] == 20180109)
]
filtered_checkins

membership_id  check_in_date  check_in_time  check_out_time
2700         48Z7A       20180109           1600            1730
2701         48Z55       20180109           1530            1700

ผู้ต้องสงสัย
1. Member id :48Z7A

   person id : 28819

   Name : Joe Germuska

2. Member id : 48Z55
   
   person id : 67318
   
   Name : Jeremy Bowers


In [130]:
murderer_car = drivers_license[drivers_license['plate_number'].str.contains('H42W')]
murderer_car

id  age  height eye_color hair_color  gender plate_number  \
915   183779   21      65      blue     blonde  female       H42W0X   
3529  423327   30      70     brown      brown    male       0H42W2   
6240  664760   21      71     black      black    male       4H42WR   

       car_make car_model  
915      Toyota     Prius  
3529  Chevrolet  Spark LS  
6240     Nissan    Altima

In [131]:
murderer = person_df[person_df['license_id'].isin([183779, 423327])]
murderer

id            name  license_id  address_number  \
6327  67318   Jeremy Bowers      423327             530   
7578  78193  Maxine Whitely      183779             110   

        address_street_name        ssn  
6327  Washington Pl, Apt 3A  871539279  
7578                Fisk Rd  137882671

# สรุปฆาตกร

จากข้อมูล
- มีกระเป๋าของ "Get Fit Now Gym"
- หมายเลขสมาชิกบนกระเป๋าขึ้นต้นด้วย "48Z"
- เฉพาะสมาชิกระดับ Gold เท่านั้นที่มีกระเป๋าแบบนี้
- ชายคนนั้นขึ้นรถยนต์ที่มีป้ายทะเบียนมีส่วนประกอบของ "H42W"
- วันที่ 9 มกราคม เขาออกไปฟิตเนส

สามารถสรุปตัวคนร้ายได้ คือ

**Name : Jeremy Bowers**

person id : 67318

## ตรวจสอบคำให้การคนร้าย

In [132]:
interview_Jeremy = interreview_df[interreview_df['person_id'] == 67318]

if not interview_Jeremy.empty:
    for index, row in interview_Jeremy.iterrows():
        print(f"--- Row Index: {index} ---")
        for column, value in row.items():
            print(f"{column}: {value}")
else:
    print("DataFrame 'interview_Jeremy' is empty.")

--- Row Index: 4990 ---
person_id: 67318
transcript: I was hired by a woman with a lot of money. I don't know her name but I know she's around 5'5" (65") or 5'7" (67"). She has red hair and she drives a Tesla Model S. I know that she attended the SQL Symphony Concert 3 times in December 2017.



### Jeremy ให้การ
person_id: 67318

transcript : ฉันถูกว่าจ้างโดยผู้หญิงคนหนึ่งที่มีเงินมาก
ฉันไม่รู้จักชื่อของเธอ แต่ฉันรู้ว่าเธอสูงประมาณ 5 ฟุต 5 นิ้ว (65 นิ้ว) หรือ 5 ฟุต 7 นิ้ว (67 นิ้ว)
เธอมีผมสีแดง และขับรถเทสล่า รุ่น Model S
ฉันรู้ว่าเธอเคยไปงาน SQL Symphony Concert 3 ครั้งในเดือนธันวาคม ปี 2017


# หาคนบงการ
- ผู้หญิงมีเงินมาก
- เธอสูงประมาณ 5 ฟุต 5 นิ้ว (65 นิ้ว) หรือ 5 ฟุต 7 นิ้ว (67 นิ้ว)
- มีผมสีแดง
- ขับรถเทสล่า รุ่น Model S
- เธอเคยไปงาน SQL Symphony Concert 3 ครั้งในเดือนธันวาคม ปี 2017

## ตรวจสอบรถ

In [133]:
path_to_drivers_license = '/content/drive/MyDrive/Python&DataViz2025/Midterm/drivers_license.csv'

In [134]:
drivers_license = pd.read_csv(path_to_drivers_license)

In [135]:
mastermind_car = drivers_license[
    (drivers_license['hair_color'] == 'red') &
    (drivers_license['gender'] == 'female') &
    (drivers_license['car_make'] == 'Tesla') &
    (drivers_license['car_model'] == 'Model S')
]
mastermind_car

id  age  height eye_color hair_color  gender plate_number car_make  \
1105  202298   68      66     green        red  female       500123    Tesla   
2054  291182   65      66      blue        red  female       08CM64    Tesla   
9078  918773   48      65     black        red  female       917UU3    Tesla   

     car_model  
1105   Model S  
2054   Model S  
9078   Model S

## หาไอดี เทียบ เลขใบขับขี่

In [136]:
mastermind1 = person_df[person_df['license_id'] == 202298]
mastermind1

id              name  license_id  address_number address_street_name  \
9985  99716  Miranda Priestly      202298            1883          Golden Ave   

            ssn  
9985  987756388

In [137]:
mastermind2 = person_df[person_df['license_id'] == 291182]
mastermind2

id           name  license_id  address_number address_street_name  \
8977  90700  Regina George      291182             332           Maple Ave   

            ssn  
8977  337169072

In [138]:
mastermind3 = person_df[person_df['license_id'] == 	918773]
mastermind3

id      name  license_id  address_number address_street_name  \
7664  78881  Red Korb      918773             107         Camerata Dr   

            ssn  
7664  961388910

##เทียบ ssn กับรายได้

In [139]:
path_to_income = '/content/drive/MyDrive/Python&DataViz2025/Midterm/income.csv'

In [140]:
income_df = pd.read_csv(path_to_income)

In [141]:
#คนที่1
mastermind1_income = income_df[income_df['ssn'] == 987756388]
mastermind1_income

ssn  annual_income
7422  987756388         310000

In [142]:
#คนที่2
mastermind2_income = income_df[income_df['ssn'] == 3337169072]
mastermind2_income

Empty DataFrame
Columns: [ssn, annual_income]
Index: []

In [143]:
#คนที่3
mastermind3_income = income_df[income_df['ssn'] == 961388910]
mastermind3_income

ssn  annual_income
7207  961388910         278000

ผู้ต้องสงสัยคนบงการ เหลือแค่

**Mastermind1**

Name : Miranda Priestly

ssn : 987756388

**Mastermind3**

Name : Red Korb

ssn : 961388910

## ตรวจสอบ Event

In [144]:
path_to_event = '/content/drive/MyDrive/Python&DataViz2025/Midterm/facebook_event_checkin.csv'

In [145]:
event_df = pd.read_csv(path_to_event)
event_df

person_id  event_id                                         event_name  \
0          28508      5880    Nudists are people who wear one-button suits.\n   
1          63713      3865  but that's because it's the best book on anyth...   
2          63713      3999           If Murphy's Law can go wrong, it will.\n   
3          63713      6436  Old programmers never die.  They just branch t...   
4          82998      4470               Help a swallow land at Capistrano.\n   
...          ...       ...                                                ...   
20006      99716      1143                               SQL Symphony Concert   
20007      99716      1143                               SQL Symphony Concert   
20008      99716      1143                               SQL Symphony Concert   
20009      67318      4719                             The Funky Grooves Tour   
20010      67318      1143                               SQL Symphony Concert   

           date  
0      20170913  
1      20171009  
2      20170502  
3      20170926  
4      20171022  
...         ...  
20006  20171206  
20007  20171212  
20008  20171229  
20009  20180115  
20010  20171206  

[20011 rows x 4 columns]

In [146]:
#คนที่1
mastermind1_event = event_df[event_df['person_id'] == 99716]
mastermind1_event

person_id  event_id            event_name      date
20006      99716      1143  SQL Symphony Concert  20171206
20007      99716      1143  SQL Symphony Concert  20171212
20008      99716      1143  SQL Symphony Concert  20171229

In [147]:
#คนที่3
mastermind3_event = event_df[event_df['person_id'] == 78881]
mastermind3_event

Empty DataFrame
Columns: [person_id, event_id, event_name, date]
Index: []

#สรุปคนบงการ

จากข้อมูลหลักฐาน ได้แก่
- ผู้หญิงมีเงินมาก
- เธอสูงประมาณ 5 ฟุต 5 นิ้ว (65 นิ้ว) หรือ 5 ฟุต 7 นิ้ว (67 นิ้ว)
- มีผมสีแดง
- ขับรถเทสล่า รุ่น Model S
- เธอเคยไปงาน SQL Symphony Concert 3 ครั้งในเดือนธันวาคม ปี 2017

สามารถสรุปผู้บงการการฆาตกรรมครั้งนี้ คือ

Name : Miranda Priestly

Id : 99716

# ตอบคำถาม
 ให้สืบหาฆาตรกร (10 คะแนน) และ ผู้บงการ (5 คะแนน) ในคดี ฆาตรกรรม (murder) ที่เกิดขึ้นในวันที่ 15 มกราคม 2018 ที่เมือง Pandas City โดยเริ่มจากการตรวจสอบ ข้อมูลคดีในไฟล์ crime_scene_report.csv


▶ **Murderer** 🔪🔪🔪

Name : Jeremy Bowers

person id : 67318

▶ **Mindmaster** 🧠🧠🧠

Name : Miranda Priestly

Id : 99716
